### ISSUE TITLE
Deprecate DEFAULT_API_VERSION

### ISSUE THREAD

**Issue by demianbrecht** [simple-salesforce/simple-salesforce#115]
This is more of a poll at this point than anything. The existence of `DEFAULT_API_VERSION` in conjunction with the optional kwarg in `Salesforce` and `SFType` seem to be causing more trouble than it's worth. New adopters want to use the latest version (as they should). If `DEFAULT_API_VERSION` is changed, you risk breaking existing code if there have been any backwards incompatible changes made across all of the services exposed by simple-salesforce.

My proposal is to deprecate `DEFAULT_API_VERSION` and related optional kwargs and instead make them required arguments. Yes, the proposed change breaks backwards compatibility on two levels, but it's to alleviate future confusion and pain around this.

Would love to hear feedback on this prior to going ahead with the change.

**Comment by demianbrecht** [comment]
@nickcatal

**Comment by nickcatal** [comment]
Are you thinking we depreciate using kwargs altogether in `Salesforce.__init__()`?

I agree that it's out of control, but I'm trying to think of a better solution.

**Comment by demianbrecht** [comment]
Yeah, I'm thinking of just adding it to the required params in `Salesforce.__init__()` and adding constants for each version (i.e. `SPRING_16 = '36.0'`). I think it's a better solution than what we have right now because it's explicit, obvious and clients are responsible for updating the versions themselves. The potential for breaking existing code is removed from simple-salesforce altogether, which is optimal. The importance of this will only increase as more service implementations are added.

**Comment by nickcatal** [comment]
I'm mostly wondering how many people ever care about using anything other than the latest version of the API.

Making it a required attribute makes sense, it just seems like trying to remember or having to lookup what the latest version number is kinda makes things more difficult than it needs to be for most use cases.

**Comment by demianbrecht** [comment]
> I'm mostly wondering how many people ever care about using anything other than the latest version of the API.
I'm not sure if there's any official policy around backwards compatibility on API releases. Due to that, there's a *possibility* of breaking existing code with an update to the API version. If users are using un-pinned dependencies, there's a distinct possibility of breaking running code with a simple-salesforce release, which shouldn't ever happen. I'm proposing break once in order to avoid any future breaks.

> Making it a required attribute makes sense, it just seems like trying to remember or having to lookup what the latest version number is kinda makes things more difficult than it needs to be for most use cases.
Lookup should be facilitated with a combination of constants and documentation updates pointing users to the latest release.

**Comment by mitchmcc** [comment]
I would vote for either having it default to the latest version, or else make it always required.  I am having trouble as I need to get past the default version provided, "29.0", and am missing something.  If I use SalesforceLogin I can pass the sf_version, but it does not return the "sf" object that is used in the other API calls.   If I use SalesforceLogin, how do I get the "sf" object returned?  Do I need to use both SalesforceLogin and Salesforce together?

**Comment by espoelstra** [comment]
@mitchmcc With the current version of `Salesforce()` you simply pass `Salesforce(<args>, version=DESIRED_API_VERSION)` and it handles getting it to SalesforceLogin.

To the topic at hand, I agree it would be better to set `version=None` in the `__init__` and fail with an error about it being a required field rather than having a default that is never up to date and causes many questions/issues. I can think of half a dozen I've seen recently related to the API version being old causing skewed results in queries.

**Comment by SUNNUWORKS** [comment]
I really do upvote this topic, as sometimes while working with the API, we need to extract data or something else and some of the salesforce objects are available in API version 46 or higher than that, where it fails. 
Also, I don't know where I am lagging, but when I passed `version=50.0` in the `Salesforce()` along with login credentials, I wasn't able to create a job (create_query_job()) and the script failed with the following error.
```
salesforce_bulk.salesforce_bulk.BulkApiError: [400] Bulk API HTTP Error result: <?xml version="1.0" encoding="UTF-8"?><error        
   xmlns="http://www.force.com/2009/06/asyncapi/dataload">
 <exceptionCode>InvalidJob</exceptionCode>
 <exceptionMessage>Unable to find object: ColorDefinition</exceptionMessage>
```
The describe() call works fine, but I wasn't able to query the object nor do I was able to create a job.
@espoelstra @demianbrecht @nickcatal , could anyone please suggest.
